vlog4j-parser/src/test/java/org/semanticweb/vlog4j/syntax/parser/RuleParserDataSourceTest.java

```diff
@@ -0,0 +1,144 @@
+	private static final String EXAMPLE_CSV_FILE_PATH = "src/main/data/input/example.csv";
+	private static final String WIKIDATA_SPARQL_ENDPOINT_URI = "https://query.wikidata.org/sparql";
+
+	@Test
+	public void testCsvSource() throws ParsingException, IOException {
+		String input = "@source p[2] : load-csv(\"" +  EXAMPLE_CSV_FILE_PATH + "\") .";
+		CsvFileDataSource csvds = new CsvFileDataSource(new File(EXAMPLE_CSV_FILE_PATH));
+		assertEquals(csvds, RuleParser.parseDataSourceDeclaration(input));
+	}
+
+	@Test
+	public void testRdfSource() throws ParsingException, IOException {
+		String input = "@source p[3] : load-rdf(\"" + EXAMPLE_RDF_FILE_PATH + "\") .";
+		RdfFileDataSource rdfds = new RdfFileDataSource(new File(EXAMPLE_RDF_FILE_PATH));
+		assertEquals(rdfds, RuleParser.parseDataSourceDeclaration(input));
+	}
+
+	@Test(expected = ParsingException.class)
+	public void testRdfSourceInvalidArity() throws ParsingException, IOException {
+		String input = "@source p[2] : load-rdf(\"" + EXAMPLE_RDF_FILE_PATH + "\") .";
+		RuleParser.parseDataSourceDeclaration(input);
+	}
+
+	@Test
+	public void testSparqlSource() throws ParsingException, MalformedURLException {
+		String input = "@source p[2] : sparql(<" + WIKIDATA_SPARQL_ENDPOINT_URI + ">,\"disease, doid\",\"?disease wdt:P699 ?doid .\") .";
+		SparqlQueryResultDataSource sparqlds = new SparqlQueryResultDataSource(
+				new URL(WIKIDATA_SPARQL_ENDPOINT_URI), "disease, doid", "?disease wdt:P699 ?doid .");
+		assertEquals(sparqlds, RuleParser.parseDataSourceDeclaration(input));
+	}
+
+	@Test(expected = ParsingException.class)
+	public void testSparqlSourceMalformedUrl() throws ParsingException, MalformedURLException {
+		String input = "@source p[2] : sparql(<not a URL>,\"disease, doid\",\"?disease wdt:P699 ?doid .\") .";
+		RuleParser.parseDataSourceDeclaration(input);
+	}
+
+	@Test(expected = ParsingException.class)
+	public void testUnknownDataSource() throws ParsingException {
+		String input = "@source p[2] : unknown-data-source(\"hello, world\") .";
+		RuleParser.parseDataSourceDeclaration(input);
+	}
+
+	@Test
+	public void testCustomDataSource() throws ParsingException {
+		CsvFileDataSource source = mock(CsvFileDataSource.class);
+		DataSourceDeclarationHandler handler = mock(DataSourceDeclarationHandler.class);
+		ParserConfiguration parserConfiguration = new ParserConfiguration();
+		parserConfiguration.registerDataSource("mock-source", handler);
+		doReturn(source).when(handler).handleDeclaration(ArgumentMatchers.<List<String>>any(),
+				ArgumentMatchers.<SubParserFactory>any());
+
+		String input = "@source p[2] : mock-source(\"hello\", \"world\") .";
+		List<String> expectedArguments = Arrays.asList("hello", "world");
+		RuleParser.parseDataSourceDeclaration(input, parserConfiguration);
+
+		verify(handler).handleDeclaration(eq(expectedArguments), ArgumentMatchers.<SubParserFactory>any());
+	}
+
```

Comment:
it is disputable, but these are more roudtripping tests testing toString then parsing tests.

Reply:
Not at all, this test case does not involve toString, nor round-tripping; it verifies that a `DataSourceDeclarationHandler` is invoked with the correct arguments.